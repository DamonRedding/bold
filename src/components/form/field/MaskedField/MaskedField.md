# MaskedField

A prop 'mask' é um array de strings ou expressões regulares, sendo as strings os caracteres fixos da máscara e as expressões regulares a expressão que pode ser digitada pelo usuário nesta posição.

No caso do telefone, por exemplo:

```
mask={['(', /\d/, /\d/, ')', ' ', /\d/, /\d/, /\d/, /\d/, /\d/, '-', /\d/, /\d/, /\d/, /\d/]}
```

Componente construído a partir de https://github.com/text-mask/text-mask.
