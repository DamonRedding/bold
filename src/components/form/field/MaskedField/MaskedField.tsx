import * as React from 'react'

import { FormFieldProps } from '../../../form/FormField'
import { Field, FieldProps, RenderProps } from '../../finalForm/Field'
import { MaskedInput, MaskedInputProps } from '../../input/MaskedInput/MaskedInput'

export interface MaskedFieldProps extends FormFieldProps, MaskedInputProps,
    Pick<FieldProps, 'parse' | 'format'> {
    name: string
}

export class MaskedField extends React.Component<MaskedFieldProps> {

    render() {
        return (
            <Field
                {...this.props}
                render={this.renderInput}
            />
        )
    }

    private renderInput = (props: RenderProps) => {
        const { format, parse, ...rest } = this.props
        return (
            <MaskedInput
                {...rest}
                {...props.input}
                status={props.hasError && 'error'}
            />
        )
    }

}
