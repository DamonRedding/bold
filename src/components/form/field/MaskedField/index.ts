export { MaskedField, MaskedFieldProps } from './MaskedField'
export * from './maskedFields'
