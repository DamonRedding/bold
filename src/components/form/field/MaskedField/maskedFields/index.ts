export * from './CepField'
export * from './CnsField'
export * from './CpfField'
export * from './TelefoneField'
export * from './TimeField'
