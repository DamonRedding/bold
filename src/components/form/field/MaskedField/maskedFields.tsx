import * as moment from 'moment'
import * as React from 'react'
import { Omit } from 'react-redux'
import createAutoCorrectedDatePipe from 'text-mask-addons/dist/createAutoCorrectedDatePipe'

import { masks, onlyNumbers } from '../../../../util/masks'
import { MaskType } from '../../input/MaskedInput/MaskedInput'
import { MaskedField, MaskedFieldProps } from '../MaskedField/MaskedField'

export interface GenericMaskedProps extends Omit<MaskedFieldProps, 'mask'> {
}

const createMaskedField = (displayName: string, mask: MaskType, props?: Partial<GenericMaskedProps>) => {
    return class extends React.PureComponent<GenericMaskedProps> {
        static displayName = displayName

        render() {
            return (
                <MaskedField mask={mask} {...props} {...this.props} />
            )
        }
    }
}

export const TelefoneField = createMaskedField('TelefoneField', masks.telefone, {
    parse: onlyNumbers,
    placeholder: '(__) _____-____',
})
export const CpfField = createMaskedField('CpfField', masks.cpf, {
    parse: onlyNumbers,
    placeholder: '___.___.___-__',
})
export const CepField = createMaskedField('CepField', masks.cep, {
    parse: onlyNumbers,
    placeholder: '_____-___',
})
export const TimeField = createMaskedField('TimeField', masks.time, {
    placeholder: 'hh:mm',
    pipe: createAutoCorrectedDatePipe('HH:MM'),
    format: (value: string) => {
        if (value) {
            return moment(value, 'hh:mm').format('hh:mm')
        }
        return null
    },
})
export const CnsField = createMaskedField('CnsField', masks.cns, {
    parse: onlyNumbers,
})
