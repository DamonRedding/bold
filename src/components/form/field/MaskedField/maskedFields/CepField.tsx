import React from 'react'

import { masks, onlyNumbers } from '../../../../../util/masks'
import { MaskedField, MaskedFieldProps } from '../../MaskedField'

export interface CepFieldProps extends MaskedFieldProps {}

export class CepField extends React.Component<CepFieldProps> {
  static defaultProps: Partial<CepFieldProps> = {}

  render() {
    return <MaskedField mask={masks.cep} placeholder='_____-___' parse={onlyNumbers} {...this.props} />
  }
}
