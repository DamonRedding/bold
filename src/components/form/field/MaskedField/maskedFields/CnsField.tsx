import React from 'react'

import { masks, onlyNumbers } from '../../../../../util/masks'
import { MaskedField, MaskedFieldProps } from '../MaskedField'

export interface CnsFieldProps extends MaskedFieldProps {

}

export class CnsField extends React.Component<CnsFieldProps> {

    static defaultProps: Partial<CnsFieldProps> = {}

    render() {
        return (
            <MaskedField
                mask={masks.cns}
                parse={onlyNumbers}
                {...this.props}
            />
        )
    }

}
