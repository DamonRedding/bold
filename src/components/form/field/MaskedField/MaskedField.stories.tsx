import { action } from '@storybook/addon-actions'
import { boolean, withKnobs } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import * as React from 'react'

import { withForm } from '../../../../stories-addons/withForm'
import { withPropTypes } from '../../../../stories-addons/withPropTypes'
import { withTheme } from '../../../../stories-addons/withTheme'
import { VFlow } from '../../../layout/Flow/VFlow'

import { MaskedField } from './MaskedField'
import * as Doc from './MaskedField.md'
import { CnsField, CpfField, TelefoneField, TimeField } from './maskedFields'

storiesOf('Form/Fields', module)
    .addDecorator(withPropTypes(Doc))
    .addDecorator(withKnobs)
    .addDecorator(withTheme())
    .addDecorator(withForm())
    .add('MaskedField', () => (
        <VFlow>
            <MaskedField
                mask={['(', /\d/, /\d/, /\d/, ')']}
                name='generico'
                label='Masked Field GenÃ©rico'
                disabled={boolean('disabled', false)}
                guide={boolean('guide', true)}
                keepCharPositions={boolean('keepCharPositions', false)}
            />
            <CpfField
                name='cpf'
                label='CPF'
                disabled={boolean('disabled', false)}
                guide={boolean('guide', true)}
                keepCharPositions={boolean('keepCharPositions', false)}
            />
            <TelefoneField
                name='telefone'
                label='Telefone'
                disabled={boolean('disabled', false)}
                guide={boolean('guide', true)}
                keepCharPositions={boolean('keepCharPositions', false)}
                icon={{ icon: 'search', onClick: action('clicked') }}
            />
            <TimeField
                name='time'
                label='Time (hour:minute)'
                disabled={boolean('disabled', false)}
                guide={boolean('guide', true)}
                keepCharPositions={boolean('keepCharPositions', false)}
            />
            <CnsField
                name='CNS'
                label='CNS'
                disabled={boolean('disabled', false)}
                guide={boolean('guide', true)}
                keepCharPositions={boolean('keepCharPositions', false)}
            />
        </VFlow>
    ))
