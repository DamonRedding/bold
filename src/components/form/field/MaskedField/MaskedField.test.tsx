import React from 'react'
import { render } from 'react-testing-library'

import { withForm } from '../../../../test/index'
import { Form } from '../../finalForm/Form/Form'

import { MaskedField } from './MaskedField'
import { TimeField } from './maskedFields'

it('render correctly', () => {
  const { container } = render(
    withForm(<MaskedField mask={['(', /\w/, ')']} label='Mask test' name='test' placeholder='Test' disabled={false} />)
  )
  expect(container).toMatchSnapshot()
})

it('render with icon', () => {
  const { container } = render(
    withForm(<MaskedField mask={['(', /\w/, ')']} name='test' icon='zoomOutline' onIconClick={jest.fn()} />)
  )
  expect(container).toMatchSnapshot()
})

it('TimeField should parse hh:mm:ss to hh:mm', () => {
  const timeField = () => {
    return <TimeField name='test' label='Test Label' />
  }
  const { container } = render(<Form render={timeField} onSubmit={null} initialValues={{ test: '05:30:55' }} />)

  expect(container.querySelector('input').getAttribute('value')).toEqual('05:30')
})
