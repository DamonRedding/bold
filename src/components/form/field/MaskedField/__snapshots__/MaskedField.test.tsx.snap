// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
.emotion-0 {
  font-weight: bold;
  display: block;
  margin-bottom: 0.25rem;
}

.emotion-1 {
  background-color: #ffffff;
  border: solid 1px #b3b3b3;
  border-radius: 2px;
  color: #4d4d4d;
  font-family: "Noto Sans",sans-serif;
  font-size: 0.75rem;
  line-height: 1rem;
  padding: calc(0.5rem - 1px) 0.5rem;
  width: 100%;
  -webkit-transition: all .2s;
  transition: all .2s;
}

.emotion-1::-webkit-input-placeholder {
  color: #b3b3b3;
}

.emotion-1::-moz-placeholder {
  color: #b3b3b3;
}

.emotion-1:-ms-input-placeholder {
  color: #b3b3b3;
}

.emotion-1::placeholder {
  color: #b3b3b3;
}

.emotion-1:disabled {
  border-color: #cccccc;
  background-color: #f2f2f7;
}

.emotion-1:not(:disabled):hover {
  border-color: #999999;
}

.emotion-1:not(:disabled):focus {
  outline: none;
  box-shadow: 0 0 0 2px #f2f2f7,0 0 0 4px #056DFF;
}

.emotion-1:not(:disabled):active {
  border-color: #056DFF;
  box-shadow: 0 2px 4px 0 rgba(0,0,0,0.09);
}

<div
  data-name="test"
>
  <label
    class="emotion-0"
    for="test"
  >
    Mask test
  </label>
  <input
    class="emotion-1"
    label="Mask test"
    name="test"
    placeholder="Test"
    value=""
  />
</div>
`;
